File: Ahorn/entities/Elevator.jl
module XaphanHelperElevator

using ..Ahorn, Maple

@mapdef Entity "XaphanHelper/Elevator" Elevator(x::Integer, y::Integer,  width::Integer=32, height::Integer=8, sprite::String="objects/XaphanHelper/Elevator", endPosition::Integer=0, canTalk::Bool=false, usableInSpeedrunMode::Bool=false, timer::Number=1.00, endAreaEntrance::Bool=false, toChapter::Integer =0, destinationRoom::String="", spawnRoomX::Integer=0, spawnRoomY::Integer=0, oneUse::Bool=false, flag::String="")

const placements = Ahorn.PlacementDict(
    "Elevator (Xaphan Helper)" => Ahorn.EntityPlacement(
        Elevator,
        "rectangle",
        Dict{String, Any}()
    ),
)

Ahorn.minimumSize(entity::Elevator) = 32, 8
Ahorn.resizable(entity::Elevator) = false, false

Ahorn.selection(entity::Elevator) = Ahorn.getEntityRectangle(entity)

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::Elevator, room::Maple.Room)
    sprite = get(entity.data, "sprite", "objects/XaphanHelper/Elevator")
    if sprite == ""
        sprite = "objects/XaphanHelper/Elevator"
    end
    Ahorn.drawSprite(ctx, "$(sprite)/elevator00.png", 16, 4)
end

end

File: Loenn/entities/Elevator.lua
local Elevator = {}

Elevator.name = "XaphanHelper/Elevator"
Elevator.depth = 8999
Elevator.fieldOrder = {"x", "y", "sprite", "canTalk", "flag", "oneUse", "timer", "endPosition", "toChapter", "destinationRoom", "spawnRoomX", "spawnRoomY", "usableInSpeedrunMode", "endAreaEntrance"}
Elevator.ignoredFields = {
    "_name", "_id", "width", "height"
}
Elevator.canResize = {false, false}
Elevator.fieldInformation = {
    endPosition = {
        fieldType = "integer",
    },
    toChapter = {
        fieldType = "integer",
    },
    spawnRoomX = {
        fieldType = "integer",
    },
    spawnRoomY = {
        fieldType = "integer",
    }
}
Elevator.placements = {
    name = "Elevator",
    data = {
        width = 32,
        height = 8,
        sprite = "objects/XaphanHelper/Elevator",
        endPosition = 0,
        canTalk = false,
        usableInSpeedrunMode = false,
        timer = 1.00,
        endAreaEntrance = false,
        toChapter = 0,
        destinationRoom = "",
        spawnRoomX = 0,
        spawnRoomY = 0,
        oneUse = false,
        flag = ""
    }
}

function Elevator.texture(room, entity)
    local sprite = entity.sprite or "objects/XaphanHelper/Elevator"
    texture = sprite .. "/elevator00"

    return texture
end

Elevator.offset = {0, 0}

return Elevator